drivers/imx6ull-gpio.c

```diff
@@ -100,3 +108,206 @@ int imx6ull_gpio_config(const char *name, uint32_t mask, unsigned flags)
+{
+	if (!gpio_valid(gp)) {
+		return -EINVAL;
+	}
+
+	if ((gp->flags & GPIO_INVERTED) != 0) {
+		active = !active;
+	}
+
+	return gpio_do_set(gp, active);
+}
+
+
+int gpio_get(const gpio_info_t *gp)
+{
+	uint32_t data[1];
+	int err;
+
+	if (!gpio_valid(gp)) {
+		return -EINVAL;
+	}
+
+	err = read(gp->fd, data, sizeof(data));
+	if (err < 0) {
+		return -errno;
+	}
+	if (err != sizeof(data)) {
+		return -EIO;
+	}
+
+	if ((gp->flags & GPIO_INVERTED) != 0) {
+		data[0] = ~data[0];
+	}
+
+	return !!(data[0] & gp->pin);
+}
+
+
+int gpio_wait(const gpio_info_t *gp, int active, time_t timeout)
+{
+	time_t when, now;
+	int val;
+
+	if (!gpio_valid(gp)) {
+		return -EINVAL;
+	}
+
+	if ((gp->flags & GPIO_INVERTED) != 0) {
+		active = !active;
+	}
+
+	if (timeout != 0) {
+		gettime(&now, NULL);
+		when = now + timeout;
+	}
+
+	for (;;) {
+		val = gpio_get(gp);  // FIXME: in gpiosrv
+		if (val < 0) {
+			return val;
+		}
+		if (!!active == !!val) {
+			return 0;
+		}
+
+		if (timeout != 0) {
+			gettime(&now, NULL);
+			if (now >= when) {
+				return -ETIME;
+			}
+			now = when - now;
+			if (now > 100000) {
+				now = 100000;
+			}
+		}
+		else {
+			now = 100000;
+		}
+
+		usleep(now);
+	}
+}
+
+
+static int gpio_close(gpio_info_t *gp)
+{
+	int err;
+
+	if (!gpio_valid(gp)) {
+		return -EINVAL;
+	}
+
+	err = close(gp->fd);
+	gp->flags = 0;
+
+	return err;
+}
+
+
+int gpio_init(gpio_info_t *gp, const char *arg, unsigned flags)
+{
+	char buf[64];
+	char *endp;
+	int err, fd = -1;
+
+	if (*arg == '-') {
+		++arg;
+		flags ^= GPIO_INVERTED;
+	}
+
+	gpio_close(gp);
+
+	gp->pin = strtoul(arg, &endp, 0);
+	if ((*endp != ',' && *endp != ':') || gp->pin >= sizeof(gp->pin) * 8) {
+		return -EINVAL;
+	}
+	gp->pin = 1u << gp->pin;
+	arg = endp + 1;
+	if (*arg == '\0') {
+		return -EINVAL;
+	}
+	if (strlen(arg) > sizeof(buf) - 6) {
+		return -EINVAL;
+	}
+
+	gp->flags = flags & ~(GPIO_ACTIVE | GPIO_INITIALIZED);
+
+	err = imx6ull_gpio_config(arg, gp->pin, gp->flags);
+	if (err < 0) {
+		return err;
+	}
+
+	snprintf(buf, sizeof(buf), "%s/port", arg);
+	gp->fd = open(buf, O_RDWR);
+	if (gp->fd < 0) {
+		err = -errno;
+		printf("gpio: can't open '%s': %s (%d)\n", buf, strerror(errno), -errno);
+		if (fd >= 0) {
+			close(fd);
+		}
+		return err;
+	}
+
+	if ((flags & GPIO_OUTPUT) != 0) {
+		err = gpio_do_set(gp, !!(flags & GPIO_ACTIVE) ^ !!(flags & GPIO_INVERTED));
+		if (err != 0) {
+			printf("gpio: WARN: can't pre-set pin: %s (%d; fd = %d)\n", strerror(-err), err, gp->fd);
+		}
+	}
+
+	fd = gp->fd;
+
+	snprintf(buf, sizeof(buf), "%s/dir", arg);
+	gp->fd = open(buf, O_RDWR);
+	if (gp->fd < 0) {
+		err = -errno;
+		printf("gpio: can't open '%s': %s (%d)\n", buf, strerror(-err), err);
+		if (fd >= 0) {
+			close(fd);
+		}
+		return err;
+	}
+
+	err = gpio_do_set(gp, flags & GPIO_OUTPUT);
+	if (err != 0) {
+		printf("gpio: can't configure pin direction: %s (%d; fd = %d)\n", strerror(-err), err, gp->fd);
+	}
+
+	close(gp->fd);
```

Comment:
Move the code from lines 300-307 inside `if` statement above. Notice that `fd >= 0` for sure at this point.